### PR TITLE
Consistent use of pointers

### PR DESCRIPTION
**Pull request by elazarg**
Always use values for Interval. Always use pointers for containers.

This may break client code that creates IPBlock using empty composite literal, but that's a brittle approach; such code should use a constructor function anyway. The break is not silent, as it quickly leads to a nil access, and is relatively easy to look for: Search for `ipblock.IPBlock{}` not preceded by a slice.

This is part of #10.

### CODE REVIEW

File: pkg/interval/intervalset.go
Comment (zivnevo) [inline pkg/interval/intervalset.go:1]: Should `NewCanonicalIntervalSet()` be renamed to `NewCanonicalSet()`?

File: pkg/ipblock/ipblock.go [hunk 1]
```diff
@@ -119,9 +121,9 @@ func (b *IPBlock) ipCount() int {
 func (b *IPBlock) Split() []*IPBlock {
 	res := make([]*IPBlock, len(b.ipRange.IntervalSet))
 	for index, ipr := range b.ipRange.IntervalSet {
-		newBlock := IPBlock{}
-		newBlock.ipRange.IntervalSet = append(newBlock.ipRange.IntervalSet, interval.Interval{Start: ipr.Start, End: ipr.End})
-		res[index] = &newBlock
+		res[index] = &IPBlock{
+			ipRange: interval.FromInterval(ipr.Start, ipr.End),
```
Comment (zivnevo) [inline pkg/ipblock/ipblock.go:125]: Maybe we should change `interval.FromInterval` to take an `Interval` object and have another function `FromRange` that takes `start` and `end` parameters.

File: pkg/interval/intervalset.go [hunk 2]
```diff
@@ -83,24 +83,24 @@ func (c *CanonicalSet) String() string {
...
 	for _, interval := range other.IntervalSet {
 		c.AddInterval(interval)
 	}
 }
 
 // Copy returns a new copy of the CanonicalSet object
-func (c *CanonicalSet) Copy() CanonicalSet {
-	return CanonicalSet{IntervalSet: append([]Interval(nil), c.IntervalSet...)}
+func (c *CanonicalSet) Copy() *CanonicalSet {
+	return &CanonicalSet{IntervalSet: append([]Interval(nil), c.IntervalSet...)}
```
Comment (zivnevo) [inline pkg/interval/intervalset.go:94]:
```suggestion
	return &CanonicalSet{IntervalSet: slices.Clone(c.IntervalSet)}
```

**Review by zivnevo — CHANGES_REQUESTED** [review]

File: pkg/ipblock/ipblock.go (hunk 1 above)
Comment (elazarg) [reply pkg/ipblock/ipblock.go:125]: I agree (or maybe add a `ToSet()` method). I don't think it should be part of this PR though.

File: pkg/interval/intervalset.go
Comment (elazarg) [reply pkg/interval/intervalset.go:1]: Yes. I'll add a small PR fixing naming (IsEqual should be Equal, etc.)

File: pkg/ipblock/ipblock.go [hunk 3]
```diff
@@ -59,34 +66,40 @@ func (b *IPBlock) toIPRangesList() []string {
...
+	res := b.Copy()
 	res.ipRange.Subtract(c.ipRange)
 	return res
 }
 
 // Union returns a new IPBlock from union of input IPBlock with this IPBlock
 func (b *IPBlock) Union(c *IPBlock) *IPBlock {
-	res := &IPBlock{}
-	res.ipRange = b.ipRange.Copy()
+	res := b.Copy()
```
Comment (zivnevo) [inline pkg/ipblock/ipblock.go:102]: If aliasing is handled in `Intersect`, it should be handled in `Union` too (and maybe in `Subtract` as well).

Comment (elazarg) [reply pkg/ipblock/ipblock.go:105]: Missed it. Fixed.

**Comment by elazarg** [comment]
(manually squashed so all the commits are signed)

**Pull request merged**